### PR TITLE
Predictive code generation

### PR DESCRIPTION
**Pull request by codemakerai-dev**
Adds a feature flag that enables the predictive code generation triggered by user actions like opening a file in the editor, saving the document or closing the file.

Currently as a consequence of any of the above actions only code generation will be triggered by default, later we can add support for automatic documentation generation and or trigger any other actions.

### CODE REVIEW

**Comment by kaihaozhao** [comment]
This feature reminds me some other code gen service like tabnine provides the ability to connect to code base, and so some pre-training.

**Review by kaihaozhao — APPROVED** [review]

File: src/predictor/predictor.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+const PredictiveGenerationCoolDown = 1000 * 30;
+
+export class Predictor {
+
+    private readonly codeMakerService: CodeMakerService;
+
+    private readonly lastPredictiveGeneration: Map<string, Date>;
+
+    constructor(codeMakerService: CodeMakerService) {
```
Comment (kaihaozhao) [inline src/predictor/predictor.ts:16]: I didn't find where this constructor is called, should we update in the extension.ts? will there be another commit?

Comment (codemakerai-dev) [reply src/predictor/predictor.ts:16]: I updated the code to glue everything together.

**Comment by mcgG** [comment]
Seems like there has some conflicts
